sno/working_copy.py

```diff
@@ -1159,117 +1189,105 @@ def reset(
             table_deletes.add(table)
 
         L.debug(
-            "table_inserts: %s\ntable_deletes: %s\ntable_updates %s",
+            "reset(): table_inserts: %s, table_deletes: %s, table_updates %s",
             table_inserts,
             table_deletes,
             table_updates,
         )
 
+        structural_changes = table_inserts | table_deletes
+        if track_changes_as_dirty and structural_changes:
+            # We don't yet support tracking changes as dirty if we delete, create, or rewrite an entire table.
+            structural_changes_text = '\n'.join(structural_changes)
+            raise NotNotYetImplemented(
+                "Sorry, this operation is not possible when there are structural changes."
+                f"Structural changes are affecting:\n{structural_changes_text}"
+            )
+
         # Delete old tables
         if table_deletes:
             self.drop_table(
-                target_tree_or_commit, *[src_datasets[d] for d in table_deletes]
+                target_tree_or_commit, *[base_datasets[d] for d in table_deletes]
             )
         # Write new tables
         if table_inserts:
             # Note: write_full doesn't work if called from within an existing db session.
             self.write_full(
-                target_tree_or_commit, *[dest_datasets[d] for d in table_inserts]
+                target_tree_or_commit, *[target_datasets[d] for d in table_inserts]
             )
 
         with self.session(bulk=1) as db:
             dbcur = db.cursor()
 
             for table in table_updates:
-                src_ds = src_datasets[table]
-                dest_ds = dest_datasets[table]
+                base_ds = base_datasets[table]
+                target_ds = target_datasets[table]
                 self._update_table(
-                    commit,
-                    src_ds,
-                    dest_ds,
+                    base_ds,
+                    target_ds,
                     db,
                     dbcur,
-                    track_changes_as_dirty=not update_meta,
+                    commit,
+                    track_changes_as_dirty=track_changes_as_dirty,
                 )
 
-            if update_meta:
+            if not track_changes_as_dirty:
                 # update the tree id
                 dbcur.execute(
                     f"UPDATE {self.META_TABLE} SET value=? WHERE table_name='*' AND key='tree';",
                     (target_tree.hex,),
                 )
 
+    def _filter_by_paths(self, datasets, paths):
+        """Filters the datasets so that only those matching the paths are returned."""
+        # FIXME - looks like this only keeps things that match *all* paths, instead of *any* paths.
+        if paths:
+            for path in paths:
+                datasets = [
+                    ds for ds in datasets if os.path.commonpath([ds.path, path]) == path
```

Comment:
is this just the same as 
```suggestion
                    ds for ds in datasets if ds.path.startswith(path)
```

Reply:
in fact since startswith takes a tuple, isn't this whole method just 

```python
     def _filter_by_paths(self, datasets, paths): 
         paths = tuple(paths)
         return [ds for ds in datasets if ds.path.startswith(paths)]
```

Reply:
Thanks! Done